src/april_tag_node.cpp

```diff
@@ -72,28 +73,60 @@ class AprilTagNode
+    double tag_size_cm;
+    if(!nh.getParam("tag_size_cm", tag_size_cm))
+    {
+      ROS_ERROR("Parameter 'tag_size_cm' is required but not set");
+      exit(1);
+    }
+    tag_size_m_ = tag_size_cm / 100.0;
 
-    // Use a private node handle so that multiple instances of the node can
-    // be run simultaneously while using different parameters.
-    ros::NodeHandle private_node_handle("~"); 
-    private_node_handle.param<bool>("show_debug_image", show_debug_image_, false);
+    ros::NodeHandle private_node_handle("~");
 
-    private_node_handle.param<double>("focal_length_x_px", camera_focal_length_x_px_, 700.0); // late 2013 macbookpro retina fx = 700
-    private_node_handle.param<double>("focal_length_y_px", camera_focal_length_y_px_, 700.0);
-    private_node_handle.param<double>("principal_point_x_px_", camera_principal_point_x_px_, 320.0); // 640x480 image = (320, 240) middle
-    private_node_handle.param<double>("principal_point_y_px_", camera_principal_point_y_px_, 240.0);
+    std::string camera_ns;
+    if(!private_node_handle.getParam("camera_ns", camera_ns))
+    {
+      ROS_ERROR("Parameter 'camera_ns' is required but not set");
+      exit(1);
+    }
 
-    double tag_size_cm;
-    nh.param<double>("tag_size_cm", tag_size_cm, 5.9);
-    tag_size_m_ = tag_size_cm / 100.0;
+    private_node_handle.param<double>("focal_length_x_px", camera_focal_length_x_px_, -1.);
+    private_node_handle.param<double>("focal_length_y_px", camera_focal_length_y_px_, -1.);
+    private_node_handle.param<double>("principal_point_x_px_", camera_principal_point_x_px_, -1.);
+    private_node_handle.param<double>("principal_point_y_px_", camera_principal_point_y_px_, -1.);
 
+    if(camera_focal_length_x_px_ < 0. || camera_focal_length_y_px_ < 0.
+       || camera_principal_point_x_px_ < 0. || camera_principal_point_y_px_ < 0.)
+    {
+      ROS_INFO("Camera intrinsics not supplied. Waiting to fetch from topic '%s/camera_info'", camera_ns.c_str());
+      auto camera_info = ros::topic::waitForMessage<sensor_msgs::CameraInfo>(camera_ns + "/camera_info", nh);
+      if(!camera_info)
+      {
+        ROS_INFO("Interrupted by node shutdown.");
+        exit(0);
+      }
+      /* sensor_msgs/CameraInfo only describes contents of intrinsic camera matrix in comments in msg file, so
+         there is no point having named constants here, sorry. */
+      if(std::all_of(camera_info->K.begin(), camera_info->K.end(), [](double k) { return k == 0;}))
+      {
+        ROS_WARN("Intrinsics of CameraInfo fetched from topic '%s/camera_info' are all zero."
+                 " Make sure your camera publishes the correct intrinsics", camera_ns.c_str());
+      }
+      camera_focal_length_x_px_ = camera_info->K[0];
+      camera_focal_length_y_px_ = camera_info->K[4];
+      camera_principal_point_x_px_ = camera_info->K[2];
+      camera_principal_point_y_px_ = camera_info->K[5];
+    }
 
+    ROS_INFO_STREAM("got tag size " << tag_size_m_ << "m");
```

Comment:
Why print tag_size? Unlike camerainfo, which you retrieve from a topic, this will be given by the user anyway and they don't need to see it again.